File: library/src/main/java/com/github/paolorotolo/appintro/util/CustomFontCache.java
package com.github.paolorotolo.appintro.util;

import android.content.Context;
import android.graphics.Typeface;

import java.util.Hashtable;

/**
 * Created by ameykshirsagar on 23/07/16.
 * Custom Font Cache Implementation
 * Prevent(s) memory leaks due to Typeface objects
 */
public class CustomFontCache {
    private static final String TAG = LogHelper.makeLogTag(CustomFontCache.class);

    private static final Hashtable<String, Typeface> fCache = new Hashtable<>();

    public static Typeface get(String tfn, Context ctx) {
        Typeface tf = fCache.get(tfn);
        if (tf == null) {
            try {
                tf = Typeface.createFromAsset(ctx.getAssets(), tfn);
                if (tf != null) {
                    fCache.put(tfn, tf);
                }

                return tf;
            } catch (Exception e) {
                if (("").equals(tfn)) {
                    LogHelper.w(TAG, e, "Empty path");
                } else {
                    LogHelper.w(TAG, e, tfn);
                }

                return null;
            }
        } else {
            return tf;
        }
    }
}

File: appintro/src/main/java/com/github/paolorotolo/appintro/internal/CustomFontCache.kt
package com.github.paolorotolo.appintro.internal

import android.content.Context
import android.graphics.Typeface

/**
 * Custom Font Cache Implementation.
 * Prevent(s) memory leaks due to Typeface objects
 */
internal object CustomFontCache {

    private val TAG = LogHelper.makeLogTag(CustomFontCache::class)
    private val cache = hashMapOf<String, Typeface>()

    operator fun get(path: String?, ctx: Context): Typeface? {
        if (path.isNullOrEmpty()) {
            LogHelper.w(TAG, "Empty typeface path provided!")
            return null
        }

        val storedTypeface = cache[path]
        return if (storedTypeface != null) {
            // Cache hit! Return the typeface.
            storedTypeface
        } else {
            // Cache miss! Create the typeface and store it.
            val newTypeface = Typeface.createFromAsset(ctx.assets, path)
            cache[path!!] = newTypeface
            newTypeface
        }
    }
}
